File: converter/src/main/java/net/chrissearle/converter/Convert.java
package net.chrissearle.converter;

import com.google.common.io.LittleEndianDataOutputStream;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileOutputStream;


public class Convert {
    public static void main(String[] args) {
        String pngFile = args[0];
        String datFile = args[1];

        try {
            BufferedImage image = ImageIO.read(new File(pngFile));

            try (
                    LittleEndianDataOutputStream outputStream = new LittleEndianDataOutputStream(new FileOutputStream(datFile))
            ) {

                for (int y = 0; y < image.getHeight(); y++) {
                    for (int x = 0; x < image.getWidth(); x++) {
                        int pixel = image.getRGB(x, y);
                        short rgb565 = (short) (((pixel & 0xf80000) >> 8) | ((pixel & 0xfc00) >> 5) | ((pixel & 0xf8) >> 3));
                        outputStream.writeShort(rgb565);
                    }
                }
            }
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
    }
}

File: converter/src/main/java/net/chrissearle/converter/Convert.kt
package net.chrissearle.converter

import com.google.common.io.LittleEndianDataOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStream

@Suppress("MagicNumber")
fun LittleEndianDataOutputStream.writePixelAsRGB565(pixel: Int) =
    this.writeShort(
        ((pixel and 0xf80000 shr 8) or (pixel and 0xfc00 shr 5) or (pixel and 0xf8 shr 3)),
    )

/** Writes [image] as little-endian RGB565, the format the badge display expects. */
fun writeRgb565(
    image: Image,
    output: OutputStream,
) = LittleEndianDataOutputStream(output).use { outputStream ->
    for (y in 0 until image.height) {
        for (x in 0 until image.width) {
            val pixel = image.rgb(x, y)

            outputStream.writePixelAsRGB565(pixel)
        }
    }
}

@Throws
fun main(args: Array<String>) {
    val (pngFile, datFile) = args

    val image = readPng(File(pngFile))

    FileOutputStream(datFile).use { output -> writeRgb565(image, output) }
}
